sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java

```diff
@@ -0,0 +1,184 @@
+
+  private static transient BrokerService broker;
+
+  @BeforeClass
+  public static void startBroker() throws Exception {
+    broker = new BrokerService();
+    broker.setUseJmx(false);
+    broker.setPersistenceAdapter(new MemoryPersistenceAdapter());
+    broker.addConnector(new URI("mqtt://localhost:11883"));
+    broker.start();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<byte[]> output = pipeline.apply(
+        MqttIO.read()
+            .withMqttConnectionConfiguration(
+                MqttIO.MqttConnectionConfiguration.create(
+                    "tcp://localhost:11883",
+                    "BEAM_PIPELINE",
+                    "READ_TOPIC"))
+          .withMaxNumRecords(10));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<byte[]>globally()))
+        .isEqualTo(10L);
+    PAssert.that(output).satisfies(new SerializableFunction<Iterable<byte[]>, Void>() {
+      @Override
+      public Void apply(Iterable<byte[]> input) {
+        for (byte[] element : input) {
+          String inputString = new String(element);
+          Assert.assertEquals("This is a test", inputString);
+        }
+        return null;
+      }
+    });
+
+    // produce messages on the broker in another thread
+    Thread thread = new Thread() {
+      public void run() {
+        try {
+          // gives time to the pipeline to start
+          Thread.sleep(2000);
+        } catch (Exception e) {
+          // nothing to do
+        }
+        try {
+          MqttClient client = new MqttClient("tcp://localhost:11883", "publisher");
+          client.connect();
+          for (int i = 0; i < 10; i++) {
+            MqttMessage message = new MqttMessage();
+            message.setQos(0);
+            message.setRetained(false);
+            message.setPayload("This is a test".getBytes());
+            client.publish("READ_TOPIC", message);
+          }
+          client.disconnect();
+          client.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+    };
+    thread.start();
+
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    List<MqttMessage> messages = new ArrayList<>();
+    MqttClient client = receive(messages);
+
+    Pipeline pipeline = TestPipeline.create();
+
+    ArrayList<byte[]> data = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      data.add("Test".getBytes());
+    }
+    pipeline.apply(Create.of(data))
+        .apply(MqttIO.write()
+            .withMqttConnectionConfiguration(
+                MqttIO.MqttConnectionConfiguration.create(
+                    "tcp://localhost:11883",
+                    "BEAM_PIPELINE",
+                    "WRITE_TOPIC"))
+            .withRetained(true).withQoS(2));
+    pipeline.run();
+
+    Assert.assertEquals(100, messages.size());
```

Comment:
Why should messages.size() be 100 at this point? Is this code race-free? What about thread safety? (I see that access to "messages" is not protected in any way)